File: server/src/model/MyShop.java
package model;

import config.enums.ErrorLog;

import java.util.ArrayList;
import java.util.List;

public class MyShop {
    
    private int maxSlot;
    private List<ProductSale> productList;
    private long lastTimeNpcCome;
    
    
    public MyShop(int maxSlot) {
            super();
        
        this.maxSlot = maxSlot;
        this.productList = new ArrayList<>();
        for (int i = 0; i < this.maxSlot; i++){
            ProductSale product = new ProductSale(i, null, 0);
            this.productList.add(product);
        }
        this.lastTimeNpcCome = 0;
    }


    public void setMaxSlot(int maxSlot) {
        this.maxSlot = maxSlot;
    }

    public int getMaxSlot() {
        return maxSlot;
    }

    public void setProductList(List<ProductSale> productList) {
        this.productList = productList;
    }

    public List<ProductSale> getProductList() {
        return productList;
    }

    public void setLastTimeNpcCome(long lastTimeNpcCome) {
        this.lastTimeNpcCome = lastTimeNpcCome;
    }

    public long getLastTimeNpcCome() {
        return lastTimeNpcCome;
    }


////
//    public short sell(ZPUserInfo user, int slot, ProductSale product){
//        if (this.productList.size() < this.maxSlot){
//            if (user.getAsset().getQuantityOfTwoStorageByProductId(product.getProduct().getTypeItem()) >= product.getProduct().getQuantity()){
//                if (user.reduceGold(product.getPrice())){
//                    
//                    this.productList.add(product);
//                    product.setSlot(slot);
//                    user.getAsset().takeItemToStorageById(product.getProduct().getTypeItem(), product.getProduct().getQuantity());
//
////                    return true;
//                    return ErrorLog.SUCCESS.getValue();
//                }
//                return ErrorLog.ERROR_GOLD_NOT_REDUCE.getValue();
//            }
//            return ErrorLog.ERROR_STORAGE_NOT_REDUCE.getValue();
//        }
//        return ErrorLog.ERROR_OVER_MAX_SLOT.getValue();
////        return false;
//    }
    
//
    public short sell(ZPUserInfo user, int slot, StorageItem product, int price){
        int index = this.getProductIdBySlot(slot);
        if (index >= 0){
            if (user.getAsset().getQuantityOfTwoStorageByProductId(product.getTypeItem()) >= product.getQuantity()){
                
                user.getAsset().takeItemToStorageById(product.getTypeItem(), product.getQuantity());

                this.productList.get(index).updateProductSale(product, price);

                //                    return true;
                return ErrorLog.SUCCESS.getValue();
            }
            return ErrorLog.ERROR_STORAGE_NOT_REDUCE.getValue();
        }
        return ErrorLog.ERROR_OVER_MAX_SLOT.getValue();
//        return false;
    }
    
    public short buy(ZPUserInfo user, ZPUserInfo userSell, int slot){
        /*
         * inprogress
         */
        ProductSale productSale = userSell.getAsset().getMyShop().getProductBySlot(slot);
        if (productSale != null){
            if (productSale.getProduct() != null){
                if (user.reduceGold(productSale.getPrice())){
                    if (user.getAsset().addItemToStorageById(productSale.getProduct().getTypeItem(), 
                                                             productSale.getProduct().getQuantity())){
                        
                        productSale.updateProductSale(null, 0);
                        return ErrorLog.SUCCESS.getValue();
                    }
                    user.addGold(productSale.getPrice());
                }
            }
        }
        //
        return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();    //
    }
    
    
    public ProductSale getProductBySlot(int slot){
        for (int i = 0; i < this.productList.size(); i++){
            if (this.productList.get(i).getSlot() == slot){
                return this.productList.get(i);
            }
        }
        return null;
    }
    
    public int getProductIdBySlot(int slot){
        for (int i = 0; i < this.productList.size(); i++){
            if (this.productList.get(i).getSlot() == slot){
                return i;
            }
        }
        return -1;
    }
    
//      
//    public short receiveMoneyFromSoldProduct(ZPUserInfo user, int slot){
//        int index = this.getProductIdBySlot(slot);
//        if (index < 0){
//            return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();
//        }
//        
//        if (this.productList.get(index).isIsSold()){
//            user.addGold(this.productList.get(index).getPrice());
//
//            this.productList.remove(index);
//            
//            return ErrorLog.SUCCESS.getValue();
//        }
//        return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();
//    }
      
    public short receiveMoneyFromSoldProduct(ZPUserInfo user, int slot){
        int index = this.getProductIdBySlot(slot);
        if (index < 0){
            return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();
        }
        
        if (this.productList.get(index).isIsSold()){
            user.addGold(this.productList.get(index).getPrice());

            this.productList.get(index).updateProductSale(null, 0);

            return ErrorLog.SUCCESS.getValue();
        }
        return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();
    }
    
    public short cancelSell(ZPUserInfo user, int slot){
        /*
         * inprogress
         */
        ProductSale productSale = this.getProductBySlot(slot);
        if (productSale != null){
            if (productSale.getProduct() != null){
                if (user.getAsset().addItemToStorageById(productSale.getProduct().getTypeItem(), 
                                                         productSale.getProduct().getQuantity())){
                    
                    productSale.updateProductSale(null, 0);
                    return ErrorLog.SUCCESS.getValue();
                }
            }
        }
        //
        return ErrorLog.ERROR_NOT_EXIST_SLOT.getValue();    //
    }
    
    public short unlockSlot(ZPUserInfo user){
        /*
         *inprogress
         */
        if (user.reduceRuby(6)){
            this.maxSlot ++;
            
            ProductSale productSale = new ProductSale(this.maxSlot - 1, null, 0);
            this.productList.add(productSale);
            
            return ErrorLog.SUCCESS.getValue();
        }
        return ErrorLog.ERROR_RUBY_NOT_REDUCE.getValue();
    }
}

File: client/src/modules/models/MyShop.js
/**
 * Created by CPU60133_LOCAL on 12/15/2017.
 */


var MyShop = cc.Class.extend({

    maxSlot: 0,
    productList: [],
    lastTimeNpcCome: null,

    ctor: function (maxSlot, productList, lastTimeNpcCome) {

        this.maxSlot = maxSlot;
        this.productList = (productList == null) ? [] : productList;
        this.lastTimeNpcCome = lastTimeNpcCome;
    },

    // //
    // sell: function (intSlot, productSale) {
    //     if (this.productList.length < this.maxSlot){
    //         if (user.getAsset().getQuantityOfTwoStorageByProductId(productSale.product.typeItem) >= productSale.product.quantity){
    //
    //             this.productList.push(productSale);
    //             productSale.setIntSlot(intSlot);
    //
    //             user.getAsset().takeItemToStorageById(productSale.product.typeItem, productSale.product.quantity);
    //
    //             return true;
    //         }
    //     }
    //     return false;
    // },
//
    sell: function (intSlot, product, price) {
        var index = this.getProductIdBySlot(intSlot);
        if (index != null){
            if (product == null){
                return false;
            }
            if (user.getAsset().getQuantityOfTwoStorageByProductId(product.typeItem) >= product.quantity){

                user.getAsset().takeItemToStorageById(product.typeItem, product.quantity);
                this.productList[index].updateProductSale(product, price);

                return true;
            }
        }
        return false;
    },

    buy: function (userSell, intSlot) {
        // var productSale = this.getProductBySlot(intSlot);
        var productSale = userSell.getAsset().getMyShop().getProductBySlot(intSlot);    //inprogress
        //
        if (productSale != null){
            if (productSale.product != null){
                if (user.reduceGold(productSale.price)){
                    if (user.getAsset().addItemToStorageById(productSale.product.typeItem, productSale.product.quantity)){

                        productSale.updateProductSale(null, 0);
                        return true;
                    }
                    user.addGold(productSale.price);
                }
            }
        }
        return false;
    },

    getProductBySlot: function (intSlot) {
        for (var i = 0; i < this.productList.length; i++){
            if (this.productList[i].slot == intSlot){
                return this.productList[i];
            }
        }
        return null;
    },


    getProductIdBySlot: function (intSlot) {
        for (var i = 0; i < this.productList.length; i++){
            if (this.productList[i].slot == intSlot){
                return i;
            }
        }
        return null;
    },


    receiveMoneyFromSoldProduct: function (intSlot) {
        var product = this.getProductBySlot(intSlot);
        if (product == null){
            return false;
        }
        if (product.isSold){
            user.addGold(product.price);

            this.productList.slice(this.getProductIdBySlot(intSlot), 1);
            return true;
        }
        return false;
    },

    cancelSell: function (intSlot) {
        var productSale = this.getProductBySlot(intSlot);
        if (productSale != null){
            if (productSale.product != null){

                if (user.reduceRuby(1)){
                    productSale.updateProductSale(null, 0);
                    return true;
                }

                //if (user.getAsset().addItemToStorageById(productSale.product.typeItem, productSale.product.quantity)){
                //
                //    productSale.updateProductSale(null, 0);
                //    return true;
                //}
            }
        }
        return false;
    },

    unlockSlot: function () {
        if (user.reduceRuby(5)){
            this.maxSlot ++;

            var productSale = new ProductSale(this.maxSlot - 1, null, 0);
            this.productList.push(productSale);

            return true;
        }
        return false;
    }

});
